Bottom: 63ad1036c331207ccff9059308ae7df05bbc4a5a
Top:    0104729e48c06500e180c453b8b683d8e9632f91
Author: Peter Clifton <pcjc2@cam.ac.uk>
Date:   2009-11-06 16:25:02 +0000

Playing with stack rendering


---

diff --git a/src/hid/gtk/gui-output-events.c b/src/hid/gtk/gui-output-events.c
index 5065ac0..819f882 100644
--- a/src/hid/gtk/gui-output-events.c
+++ b/src/hid/gtk/gui-output-events.c
@@ -1053,7 +1053,17 @@ SetPVColor_inlayer (PinTypePtr Pin, LayerTypePtr Layer, int Type)
   else if (TEST_FLAG (FOUNDFLAG, Pin))
     color = PCB->ConnectedColor;
   else
-    color = Layer->Color;
+    {
+      int component_group = GetLayerGroupNumberByNumber (max_layer + COMPONENT_LAYER);
+      int solder_group    = GetLayerGroupNumberByNumber (max_layer + SOLDER_LAYER);
+      int this_group      = GetLayerGroupNumberByPointer (Layer);
+
+      if (this_group == component_group || this_group == solder_group)
+        color = (SWAP_IDENT == (this_group == solder_group))
+                  ? PCB->ViaColor : PCB->InvisibleObjectsColor;
+      else
+        color = Layer->Color;
+    }
 
   gui->set_color (Output.fgGC, color);
 }
@@ -1343,6 +1353,8 @@ DrawLayerGroup (int group, const BoxType * screen)
   int n_entries = PCB->LayerGroups.Number[group];
   Cardinal *layers = PCB->LayerGroups.Entries[group];
   int first_run = 1;
+  int component = GetLayerGroupNumberByNumber (max_layer + COMPONENT_LAYER);
+  int solder    = GetLayerGroupNumberByNumber (max_layer + SOLDER_LAYER);
 
   if (!gui->set_layer (0, group, 0)) {
     gui->set_layer (NULL, SL (FINISHED, 0), 0);
@@ -1400,10 +1412,18 @@ DrawLayerGroup (int group, const BoxType * screen)
         }
       }
 
-      /* Draw pins and vias on this layer */
+      /* Draw pins, vias and pads on this layer */
       if (!global_view_2d && rv) {
         if (PCB->PinOn) r_search (PCB->Data->pin_tree, screen, NULL, pin_inlayer_callback, Layer);
         if (PCB->ViaOn) r_search (PCB->Data->via_tree, screen, NULL, via_inlayer_callback, Layer);
+        if ((layernum == component && !SWAP_IDENT) ||
+            (layernum == solder    &&  SWAP_IDENT))
+          if (PCB->PinOn)
+            r_search (PCB->Data->pad_tree, screen, NULL, pad_callback, Layer);
+        if ((layernum == solder    && !SWAP_IDENT) ||
+            (layernum == component &&  SWAP_IDENT))
+          if (PCB->PinOn)
+            r_search (PCB->Data->pad_tree, screen, NULL, backPad_callback, Layer);
       }
 
       if (TEST_FLAG (CHECKPLANESFLAG, PCB))
@@ -1541,7 +1561,8 @@ ghid_draw_everything (BoxTypePtr drawn_area)
    */
   if (!TEST_FLAG (CHECKPLANESFLAG, PCB) &&
       gui->set_layer ("invisible", SL (INVISIBLE, 0), 0)) {
-    r_search (PCB->Data->pad_tree, drawn_area, NULL, backPad_callback, NULL);
+    if (global_view_2d)
+      r_search (PCB->Data->pad_tree, drawn_area, NULL, backPad_callback, NULL);
     if (PCB->ElementOn) {
       r_search (PCB->Data->element_tree, drawn_area, NULL, backE_callback, NULL);
       r_search (PCB->Data->name_tree[NAME_INDEX (PCB)], drawn_area, NULL, backN_callback, NULL);
@@ -1581,20 +1602,21 @@ ghid_draw_everything (BoxTypePtr drawn_area)
     gui->set_layer ("bottomsilk", SL (SILK, BOTTOM), 0);
 //  gui->set_layer (NULL, SL (FINISHED, 0), 0);
 
-#if 1
-  /* Mask out drilled holes */
-  hidgl_flush_triangles (&buffer);
-  glPushAttrib (GL_COLOR_BUFFER_BIT);
-  glColorMask (0, 0, 0, 0);
-  if (PCB->PinOn) r_search (PCB->Data->pin_tree, drawn_area, NULL, hole_callback, NULL);
-  if (PCB->ViaOn) r_search (PCB->Data->via_tree, drawn_area, NULL, hole_callback, NULL);
-  hidgl_flush_triangles (&buffer);
-  glPopAttrib ();
-
-  if (PCB->PinOn) r_search (PCB->Data->pad_tree, drawn_area, NULL, pad_callback, NULL);
-  if (PCB->PinOn) r_search (PCB->Data->pin_tree, drawn_area, NULL, pin_callback, NULL);
-  if (PCB->ViaOn) r_search (PCB->Data->via_tree, drawn_area, NULL, via_callback, NULL);
-#endif
+  if (global_view_2d)
+    {
+      /* Mask out drilled holes */
+      hidgl_flush_triangles (&buffer);
+      glPushAttrib (GL_COLOR_BUFFER_BIT);
+      glColorMask (0, 0, 0, 0);
+      if (PCB->PinOn) r_search (PCB->Data->pin_tree, drawn_area, NULL, hole_callback, NULL);
+      if (PCB->ViaOn) r_search (PCB->Data->via_tree, drawn_area, NULL, hole_callback, NULL);
+      hidgl_flush_triangles (&buffer);
+      glPopAttrib ();
+
+      if (PCB->PinOn) r_search (PCB->Data->pad_tree, drawn_area, NULL, pad_callback, NULL);
+      if (PCB->PinOn) r_search (PCB->Data->pin_tree, drawn_area, NULL, pin_callback, NULL);
+      if (PCB->ViaOn) r_search (PCB->Data->via_tree, drawn_area, NULL, via_callback, NULL);
+    }
 
   gui->set_layer (NULL, SL (FINISHED, 0), 0);
